search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,581 @@
+typedef struct _YelpSearchProviderApp YelpSearchProviderApp;
+
+struct _YelpSearchProviderApp {
+    GApplication parent;
+    gulong uri_resolve_id;
+    YelpShellSearchProvider2 *skeleton;
+
+    GHashTable *page_data_hash_map;
+};
+
+GType yelp_search_provider_app_get_type (void);
+
+#define YELP_TYPE_SEARCH_PROVIDER_APP yelp_search_provider_app_get_type()
+#define YELP_SEARCH_PROVIDER_APP(obj) \
+      (G_TYPE_CHECK_INSTANCE_CAST((obj), YELP_TYPE_SEARCH_PROVIDER_APP, YelpSearchProviderApp))
+
+G_DEFINE_TYPE (YelpSearchProviderApp, yelp_search_provider_app, G_TYPE_APPLICATION)
+
+static PageData *
+page_data_new_steal (gchar *title,
+                     gchar *desc,
+                     GIcon *icon)
+{
+    PageData *data = g_new0 (PageData, 1);
+    if (title) {
+        data->title = title;
+        data->title_casefold = g_utf8_casefold (title, -1);
+    }
+    if (desc) {
+        data->desc = desc;
+        data->desc_casefold = g_utf8_casefold (desc, -1);
+    }
+    data->icon = icon;
+    return data;
+}
+
+static void
+page_data_free (PageData *page_data)
+{
+    g_free (page_data->title);
+    g_free (page_data->title_casefold);
+    g_free (page_data->desc);
+    g_free (page_data->desc_casefold);
+    if (page_data->icon)
+        g_object_unref (page_data->icon);
+    g_free (page_data);
+}
+
+static GVariant *
+get_result_metas (gchar      **page_ids,
+                  GHashTable  *page_data)
+{
+    GVariantBuilder metas;
+    gint i;
+
+    g_variant_builder_init (&metas, G_VARIANT_TYPE ("aa{sv}"));
+
+    for (i = 0; page_ids[i] != NULL; i++) {
+        GVariantBuilder meta;
+        gchar *gicon_str;
+        PageData *data = g_hash_table_lookup (page_data, page_ids[i]);
+        g_variant_builder_init (&meta, G_VARIANT_TYPE ("a{sv}"));
+        g_variant_builder_add (&meta, "{sv}",
+                              "id", g_variant_new_string (page_ids[i]));
+        if (data != NULL) {
+            g_variant_builder_add (&meta, "{sv}",
+                                   "name", g_variant_new_string (data->title));
+            gicon_str = g_icon_to_string (data->icon);
+            g_variant_builder_add (&meta, "{sv}",
+                                  "gicon", g_variant_new_string (gicon_str));
+            g_free (gicon_str);
+            g_variant_builder_add (&meta, "{sv}",
+                                   "description", g_variant_new_string (data->desc));
+        } else {
+            g_variant_builder_add (&meta, "{sv}",
+                                   "name", NULL);
+            g_variant_builder_add (&meta, "{sv}",
+                                  "gicon", NULL);
+            g_variant_builder_add (&meta, "{sv}",
+                                   "description", NULL);
+        }
+        g_variant_builder_add_value (&metas, g_variant_builder_end (&meta));
+    }
+
+    return g_variant_new ("(aa{sv})", &metas);
+}
+
+// FIXME: Search was copied from yelp-search-entry and slightly adapted.
```

Comment:
These are C++ style comments, some pickier options to the compiler may cause it to bail out.